tests/test_constraint.py

```diff
@@ -0,0 +1,164 @@
+import pytest
+import numpy as np
+from numpy.testing import assert_array_equal, assert_almost_equal
+import scarlet
+
+
+class TestUpdate(object):
+
+    def test_positivity(self):
+        X = np.random.rand(100) - 0.5
+        step = 0
+
+        constraint = scarlet.PositivityConstraint()
+        X_ = constraint(X, step)
+        assert not any(X_ < 0)
+
+    def test_normalization(self):
+        X = np.random.rand(100)
+        step = 0
+
+        X_ = X.copy()
+        constraint = scarlet.NormalizationConstraint(type='sum')
+        X_ = constraint(X_, step)
+        assert_almost_equal(X_, X / X.sum())
+
+        X_ = X.copy()
+        constraint = scarlet.NormalizationConstraint(type='max')
+        X_ = constraint(X_, step)
+        assert_almost_equal(X_, X / X.max())
+
+    def test_l0(self):
+        X = np.random.rand(100) - 0.5
+        step = 0.5
+        thresh = 0.25
+
+        X_ = X.copy()
+        constraint = scarlet.L0Constraint(thresh=thresh, type='relative')
+        X_ = constraint(X_, step)
+        mask = np.abs(X) < thresh*step
+        assert all(np.abs(X_[mask]) == 0)
+        assert_array_equal(X_[~mask], X[~mask])
+
+        X_ = X.copy()
+        constraint = scarlet.L0Constraint(thresh=thresh, type='absolute')
+        X_ = constraint(X_, step)
+        mask = np.abs(X) < thresh
+        assert all(np.abs(X_[mask]) == 0)
+        assert_array_equal(X_[~mask], X[~mask])
+
+    def test_l1(self):
+        X = np.random.rand(100) - 0.5
+        step = 0.5
+        thresh = 0.25
+
+        X_ = X.copy()
+        constraint = scarlet.L1Constraint(thresh=thresh, type='relative')
+        X_ = constraint(X_, step)
+        thresh_ = thresh*step
+        mask = np.abs(X) < thresh_
+        assert all(np.abs(X_[mask]) == 0)
+        assert_array_equal(np.abs(X_[~mask]), np.abs(np.abs(X[~mask]) - thresh_))
+
+        X_ = X.copy()
+        constraint = scarlet.L1Constraint(thresh=thresh, type='absolute')
+        X_ = constraint(X_, step)
+        mask = np.abs(X) < thresh
+        assert all(np.abs(X_[mask]) == 0)
+        assert_array_equal(np.abs(X_[~mask]), np.abs(np.abs(X[~mask]) - thresh))
+
+    def test_threshold(self):
+        # Use a random seed in the test to prevent race conditions
+        np.random.seed(0)
+        noise = np.random.rand(21, 21)*2  # noise background to eliminate
+        signal = np.zeros(noise.shape)
+        func = scarlet.psf.gaussian
+        signal[7:14, 7:14] = 10 * scarlet.PSF(func, shape=(None, 21, 21)).image[0, 7:14, 7:14]
+        X = signal + noise
+
+
+        step = 0
+        constraint = scarlet.ThresholdConstraint()
+        X_ = constraint(X, step)
+
+        # regression test with thresh from reference version
+        thresh = 0.05704869232578929
+        mask = X < thresh
+        assert all(X_[mask] == 0)
+        assert_array_equal(X_[~mask], X[~mask])
+
+    def test_monotonic(self):
```

Comment:
You should also test the uncentered versions of both monotonicity and symmetry. You don't test the uncentered operation anywhere and that's a function that is prone to errors.

Reply:
We don't have uncentered versions of this constraint anymore

Reply:
What if a source is near an edge? Do you allow it to grow outside the blend frame (where we are just wasting memory) or do you truncate the unbounded side by the bounded side?

Reply:
The former. The memory overhead is secondary. There is no principled way of truncating sources on the edge once you consider the effects of the PSF. The source must be allowed to extend past the observed frame.

Reply:
What does this do for monotonicity? Symmetry I understand, since the side of the source that is within the image still has some constraints on it. But I'm not sure what will happen with only monotonicity turned on, where the source will be initialized to be symmetric and extend "out of bounds." Is there anything in the likelihood to ever restrict the source or will the out of bounds portion of the source always remain the same as it was when it was initialized?

Reply:
These "outside" pixels probably won't move much. But that's OK, as long as they are there and consistent with all data we've got. If they weren't there, it cause an underestimate of the intensity in inside pixels close to the edge.

Reply:
Ok, this is something that we should keep an eye on. At some point I can generate some fake images with sources on the edge and see what they look like with symmetry off.